tests/functional/tokens/test_generate.py

```diff
@@ -0,0 +1,231 @@
+    else:
+        # convert "['str', 'str']" to ['str', 'str']
+        payload = {"scopes": ast.literal_eval(scopes), "expires": expires}
+
+    return payload
+
+
+@when(
+    "the admin sends a POST request to '/api/v1/token/new'",
+    target_fixture="response",
+)
+def the_admin_sends_request(http_request, headers, payload):
+    response = http_request(
+        method="POST", url="/api/v1/token/new", headers=headers, json=payload
+    )
+    return response
+
+
+@then("the admin should get status code '200'")
+def the_admin_gets_status_code_200(response):
+    assert response.status_code == 200, response.text
+
+
+@then("the admin should get 'access_token' with a new token")
+def the_admin_gets_access_token(response):
+    assert response.json()["access_token"] is not None
+
+
+@scenario(
+    "../../features/tokens/generate.feature",
+    "Admin cannot generate Token using HTTP API with invalid expires",
+)
+def test_admin_cannot_generate_token_using_api_with_invalid_expires():
+    """Admin cannot generate Token using HTTP API with invalid expires."""
+
+
+@when(
+    "the admin sends a POST request to '/api/v1/token/new' with invalid "
+    "'expires' in 'payload'",
+    target_fixture="response",
+)
+def the_admin_sends_with_invalid_expires_in_payload(
+    http_request, headers, payload
+):
+    response = http_request(
+        method="POST", url="/api/v1/token/new", headers=headers, json=payload
+    )
+    return response
+
+
+@then("the admin should get status code '422'")
+def the_admin_should_get_status_code_422_invalid_expires(response):
+    assert response.status_code == 422, response.text
+
+
+@then("the admin should get not allowed scope as a given context detail")
+def the_admin_should_get_not_allowed_scope_as_a_given_context_detail(response):
+    assert response.json() == {}
+
+
+@scenario(
+    "../../features/tokens/generate.feature",
+    "Admin cannot generate Token using HTTP API for certain scopes",
+)
+def test_admin_cannot_generate_token_using_http_rest_api_for_certain_scopes():
+    """Admin cannot generate Token using HTTP API for certain scopes."""
+
+
+@when(
+    "the admin sends a POST request to '/api/v1/token/new' with not allowed "
+    "'scopes' in 'payload'",
+    target_fixture="response",
+)
+def the_admin_sends_request_with_invalid_scopes(
+    http_request, headers, payload
+):
+    response = http_request(
+        method="POST", url="/api/v1/token/new", headers=headers, json=payload
+    )
+    return response
+
+
+@then("the admin should get status code '422'")
+def the_admin_should_get_status_code_422_scopes(response):
+    assert response.status_code == 422, response.text
+
+
+@then("the admin should get not allowed scope as a given context detail")
+def the_admin_should_get_scope_in_given_context_detail(response):
+    not_allowed_scopes = ["write:token", "write:bootstrap"]
+    assert response.json()["detail"][0]["ctx"]["given"] in not_allowed_scopes
```

Comment:
Again it will be good to add one line explanation of what's happening here.
It's a little strange that you are using the same description for this step as for the step on lines 109-111, but you are doing different checks at both places.

Reply:
Same as above.